File: RxTools-library/src/main/java/com/vondear/rxtools/view/likeview/tools/RxPorterShapeImageView.java
package com.vondear.rxtools.view.likeview.tools;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;

import com.vondear.rxtools.R;

public class RxPorterShapeImageView extends RxPorterImageView {
    private Drawable shape;
    private Matrix matrix;
    private Matrix drawMatrix;

    public RxPorterShapeImageView(Context context) {
        super(context);
        setup(context, null, 0);
    }

    public RxPorterShapeImageView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setup(context, attrs, 0);
    }

    public RxPorterShapeImageView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setup(context, attrs, defStyle);
    }

    private void setup(Context context, AttributeSet attrs, int defStyle) {
        if(attrs != null){
            TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.RxPorterImageView, defStyle, 0);
            shape = typedArray.getDrawable(R.styleable.RxPorterImageView_siShape);
            typedArray.recycle();
        }
        matrix = new Matrix();
    }

    public void setShape(Drawable drawable) {
        shape = drawable;
        invalidate();
    }

    @Override
    protected void paintMaskCanvas(Canvas maskCanvas, Paint maskPaint, int width, int height) {
        if(shape != null) {
            if (shape instanceof BitmapDrawable) {
                configureBitmapBounds(getWidth(), getHeight());
                if(drawMatrix != null) {
                    int drawableSaveCount = maskCanvas.getSaveCount();
                    maskCanvas.save();
                    maskCanvas.concat(matrix);
                    shape.draw(maskCanvas);
                    maskCanvas.restoreToCount(drawableSaveCount);
                    return;
                }
            }

            shape.setBounds(0, 0, getWidth(), getHeight());
            shape.draw(maskCanvas);
        }
    }

    private void configureBitmapBounds(int viewWidth, int viewHeight) {
        drawMatrix = null;
        int drawableWidth = shape.getIntrinsicWidth();
        int drawableHeight = shape.getIntrinsicHeight();
        boolean fits = viewWidth == drawableWidth && viewHeight == drawableHeight;

        if (drawableWidth > 0 && drawableHeight > 0 && !fits) {
            shape.setBounds(0, 0, drawableWidth, drawableHeight);
            float widthRatio = (float) viewWidth / (float) drawableWidth;
            float heightRatio = (float) viewHeight / (float) drawableHeight;
            float scale = Math.min(widthRatio, heightRatio);
            float dx = (int) ((viewWidth - drawableWidth * scale) * 0.5f + 0.5f);
            float dy = (int) ((viewHeight - drawableHeight * scale) * 0.5f + 0.5f);

            matrix.setScale(scale, scale);
            matrix.postTranslate(dx, dy);
        }
    }
}

File: RxUI/src/main/java/com/tamsiree/rxui/view/likeview/tools/RxPorterShapeImageView.kt
package com.tamsiree.rxui.view.likeview.tools

import android.content.Context
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import com.tamsiree.rxui.R

/**
 * @author tamsiree
 */
open class RxPorterShapeImageView : RxPorterImageView {
    private var shape: Drawable? = null
    private var matrix0: Matrix? = null
    private var drawMatrix: Matrix? = null

    constructor(context: Context) : super(context) {
        setup(context, null, 0)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        setup(context, attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        setup(context, attrs, defStyle)
    }

    private fun setup(context: Context, attrs: AttributeSet?, defStyle: Int) {
        if (attrs != null) {
            val typedArray = context.obtainStyledAttributes(attrs, R.styleable.RxPorterImageView, defStyle, 0)
            shape = typedArray.getDrawable(R.styleable.RxPorterImageView_siShape)
            typedArray.recycle()
        }
        matrix0 = Matrix()
    }

    fun setShape(drawable: Drawable?) {
        shape = drawable
        invalidate()
    }

    override fun paintMaskCanvas(maskCanvas: Canvas?, maskPaint: Paint?, width: Int, height: Int) {
        if (shape != null) {
            if (shape is BitmapDrawable) {
                configureBitmapBounds(getWidth(), getHeight())
                if (drawMatrix != null) {
                    val drawableSaveCount = maskCanvas?.saveCount
                    maskCanvas?.save()
                    maskCanvas?.concat(matrix0)
                    shape?.draw(maskCanvas!!)
                    maskCanvas?.restoreToCount(drawableSaveCount!!)
                    return
                }
            }
            shape!!.setBounds(0, 0, getWidth(), getHeight())
            shape!!.draw(maskCanvas!!)
        }
    }

    private fun configureBitmapBounds(viewWidth: Int, viewHeight: Int) {
        drawMatrix = null
        val drawableWidth = shape!!.intrinsicWidth
        val drawableHeight = shape!!.intrinsicHeight
        val fits = viewWidth == drawableWidth && viewHeight == drawableHeight
        if (drawableWidth > 0 && drawableHeight > 0 && !fits) {
            shape!!.setBounds(0, 0, drawableWidth, drawableHeight)
            val widthRatio = viewWidth.toFloat() / drawableWidth.toFloat()
            val heightRatio = viewHeight.toFloat() / drawableHeight.toFloat()
            val scale = Math.min(widthRatio, heightRatio)
            val dx: Float = ((viewWidth - drawableWidth * scale) * 0.5f + 0.5f)
            val dy: Float = ((viewHeight - drawableHeight * scale) * 0.5f + 0.5f)
            matrix0!!.setScale(scale, scale)
            matrix0!!.postTranslate(dx, dy)
        }
    }
}